command/loadbot/execution.go

```diff
@@ -211,20 +253,83 @@ func (l *Loadbot) Run() error {
 		return fmt.Errorf("failed to extract sender account: %w", err)
 	}
 
-	client, err := createJSONRPCClient(l.cfg.JSONRPC, l.cfg.MaxConns)
+	jsonClient, err := createJSONRPCClient(l.cfg.JSONRPC, l.cfg.MaxConns)
+	if err != nil {
+		return fmt.Errorf("an error has occurred while creating JSON-RPC client: %w", err)
+	}
+
+	grpcClient, err := createGRPCClient(l.cfg.GRPC)
 	if err != nil {
 		return fmt.Errorf("an error has occurred while creating JSON-RPC client: %w", err)
 	}
 
 	defer func(client *jsonrpc.Client) {
-		_ = shutdownClient(client)
-	}(client)
+		_ = client.Close()
+	}(jsonClient)
 
-	nonce, err := getInitialSenderNonce(client, sender.Address)
+	nonce, err := getInitialSenderNonce(jsonClient, sender.Address)
 	if err != nil {
-		return fmt.Errorf("an error occurred while getting initial sender nonce: %w", err)
+		return fmt.Errorf("unable to get initial sender nonce: %w", err)
+	}
+
+	gasPrice := l.cfg.GasPrice
+	if gasPrice == nil {
+		// No gas price specified, query the network for an estimation
+		avgGasPrice, err := getAverageGasPrice(jsonClient)
+		if err != nil {
+			return fmt.Errorf("unable to get average gas price: %w", err)
+		}
+
+		gasPrice.SetUint64(avgGasPrice)
+	}
+
+	// Set up the transaction generator
+	generatorParams := &generator.GeneratorParams{
+		Nonce:         nonce,
+		ChainID:       l.cfg.ChainID,
+		SenderAddress: sender.Address,
+		SenderKey:     sender.PrivateKey,
+		GasPrice:      gasPrice,
+		Value:         l.cfg.Value,
+	}
+
+	var (
+		txnGenerator generator.TransactionGenerator
+		genErr       error = nil
+	)
+
+	switch l.cfg.GeneratorMode {
+	case transfer:
+		txnGenerator, genErr = generator.NewTransferGenerator(generatorParams)
+	case deploy:
+		txnGenerator, genErr = generator.NewDeployGenerator(generatorParams)
 	}
 
+	if genErr != nil {
+		return fmt.Errorf("unable to start generator, %w", genErr)
+	}
+
+	l.generator = txnGenerator
+
+	// Get the gas estimate
+	exampleTxn, err := l.generator.GetExampleTransaction()
+	if err != nil {
+		return fmt.Errorf("unable to get example transaction, %w", err)
+	}
+
+	gasLimit := l.cfg.GasLimit
+	if gasLimit == nil {
+		// No gas limit specified, query the network for an estimation
+		gasEstimate, estimateErr := estimateGas(jsonClient, exampleTxn)
+		if estimateErr != nil {
+			return fmt.Errorf("unable to get gas estimate, %w", err)
+		}
+
+		gasLimit.SetUint64(gasEstimate)
```

Comment:
panic happens when`gasLimit`  is nil

```suggestion
		gasLimit = new(big.Int).SetUint64(gasEstimate)
```

Reply:
Good catch 💯 

Fixed 👍 

6e01c9f